src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c

```diff
@@ -0,0 +1,1041 @@
+
+    if (!ptr)
+    {
+        return NULL;
+    }
+
+    ptr->mainloop = pa_threaded_mainloop_new();
+
+    if (!ptr->mainloop)
+    {
+        goto fail;
+    }
+
+    memset(buf, 0x00, PATH_MAX + 20);
+    snprintf(buf, sizeof(buf), "Portaudio Pulseaudio HostApi");
+
+    ptr->context =
+        pa_context_new(pa_threaded_mainloop_get_api(ptr->mainloop), buf);
+
+    if (!ptr->context)
+    {
+        goto fail;
+    }
+
+    pa_context_set_state_callback(ptr->context, PulseaudioCheckContextStateCb, ptr);
+
+
+    if (pa_threaded_mainloop_start(ptr->mainloop) < 0)
+    {
+        goto fail;
+    }
+
+    ptr->deviceCount = 0;
+
+    return ptr;
+
+fail:
+    PulseaudioFree(ptr);
+    free(ptr);
+    return NULL;
+}
+
+void PulseaudioFree(PaPulseaudioHostApiRepresentation *ptr)
+{
+    if (ptr->mainloop)
+    {
+        pa_threaded_mainloop_stop(ptr->mainloop);
+    }
+
+    if (ptr->context)
+    {
+        pa_context_disconnect(ptr->context);
+        pa_context_unref(ptr->context);
+    }
+
+    if (ptr->mainloop)
+    {
+        pa_threaded_mainloop_free(ptr->mainloop);
+    }
+
+}
+
+
+static void PulseaudioCheckContextStateCb(pa_context * c, void *userdata)
+{
+    PaPulseaudioHostApiRepresentation *ptr = userdata;
+    assert(c);
+
+    ptr->state = pa_context_get_state(c);
+    pa_threaded_mainloop_signal(ptr->mainloop, 0);
+}
+
+
+void PulseaudioSinkListCb(pa_context *c, const pa_sink_info *l, int eol, void *userdata)
+{
+    PaPulseaudioHostApiRepresentation *l_ptrHostApi = (PaPulseaudioHostApiRepresentation *)userdata;
+    PaError result = paNoError;
+    char *l_strName = NULL;
+
+    /* If eol is set to a positive number, you're at the end of the list */
+    if (eol > 0)
+    {
+        goto error;
+    }
+
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].structVersion = 2;
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].hostApi = l_ptrHostApi->hostApiIndex;
+
+    l_strName = (char *)l->name;
+
+    if(l->description != NULL)
+    {
+        l_strName = (char *)l->description;
+    }
+
+    PA_UNLESS(l_ptrHostApi->pulseaudioDeviceNames[l_ptrHostApi->deviceCount] = (char*)PaUtil_GroupAllocateMemory(l_ptrHostApi->allocations,
+              strnlen(l->name, 1024) + 1), paInsufficientMemory);
+
+    memset(l_ptrHostApi->pulseaudioDeviceNames[l_ptrHostApi->deviceCount], 0x00, strnlen(l->name, 1024) + 1);
+    strncpy((char *)l_ptrHostApi->pulseaudioDeviceNames[l_ptrHostApi->deviceCount], l->name, strnlen(l->name, 1024));
+
+    PA_UNLESS(l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].name = (char*)PaUtil_GroupAllocateMemory(l_ptrHostApi->allocations,
+              strnlen(l_strName, 1024) + 1), paInsufficientMemory);
+
+    memset((char *)l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].name, 0x00, strnlen(l_strName, 1024) + 1);
+    strncpy((char *)l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].name, l_strName, strnlen(l_strName, 1024));
```

Comment:
I don't now the PA style guide, but these lines above are not entirely displayed in github.